libzmix/src/utils/bls12_381/mod.rs

```diff
@@ -0,0 +1,237 @@
+use amcl::bls381::big::BIG;
+use amcl::bls381::dbig::DBIG;
+use amcl::bls381::ecp::ECP;
+use amcl::bls381::ecp2::ECP2;
+use amcl::bls381::rom;
+
+use sha2::Sha256;
+
+use rand::thread_rng;
+use rand_core::RngCore;
+
+use hkdf::Hkdf;
+
+lazy_static! {
+    static ref G1: ECP = ECP::generator();
+    static ref G2: ECP2 = ECP2::generator();
+    static ref GROUP_ORDER: BIG = BIG::new_ints(&rom::CURVE_ORDER);
+}
+
+macro_rules! serialize_impl {
+    ($name:ident, $internal:ident, $visitor:ident) => {
+        impl std::str::FromStr for $name {
+            type Err = std::num::ParseIntError;
+
+            fn from_str(s: &str) -> Result<Self, Self::Err> {
+                Ok($name($internal::from_hex(s.to_string())))
+            }
+        }
+
+        impl From<&str> for $name {
+            fn from(data: &str) -> $name {
+                $name($internal::from_hex(data.to_string()))
+            }
+        }
+
+        impl From<&[u8]> for $name {
+            fn from(data: &[u8]) -> $name {
+                let mut vec = data.to_vec();
+                if data.len() > Self::BYTES_REPR_SIZE {
+                    vec = data[0..Self::BYTES_REPR_SIZE].to_vec();
```

Comment:
I find it better to be explicit and fail if there are more bytes than needed than silently using a truncated version. Caller might find absurd giving 2 different byte slices which are same in first `BYTES_REPR_SIZE` bytes giving the same field element.

Reply:
Not opposed to this suggestion.